File: students/UnitySample/Assets/Scripts/BrainBitController.cs
using NeuroSDK;
using SignalMath;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using UnityEngine;

public enum ResistState
{
    Normal, Bad
}
public struct ResistValues
{
    public ResistState O1;
    public ResistState O2;
    public ResistState T3;
    public ResistState T4;
}

public class MindDataReal
{
    public double Attention;
    public double Relaxation;
}

public class BrainBitInfo
{
    public string Name { get; set; }
    public string Address { get; set; }
    public SensorInfo SensInfo { get; set; }
    public ConnectionState ConnectionState { get; set; }
}

public enum ConnectionState
{
    Connection,
    Connected,
    Disconnection,
    Disconnected,
    Error
}

class BrainBitAdditional
{
    public bool needReconnect = false;
    public BrainBitSensor bb;
    public EegEmotionalMath emotionalMath;
    public bool isSignal = false;
    public bool calibrationStarted = false;

    public BrainBitAdditional(bool needReconnect, BrainBitSensor bb)
    {
        this.needReconnect = needReconnect;
        this.bb = bb;
        emotionalMath = CreateEmotionalMath();
    }

    private EegEmotionalMath CreateEmotionalMath()
    {
        int samplingFrequencyHz = 250;
        var mathLib = new MathLibSetting
        {
            sampling_rate = (uint)samplingFrequencyHz,
            process_win_freq = 25,
            fft_window = (uint)samplingFrequencyHz * 2,
            n_first_sec_skipped = 4,
            bipolar_mode = true,
            squared_spectrum = true,
            channels_number = (uint)1,
            channel_for_analysis = 0
        };

        var artsDetect = new ArtifactDetectSetting
        {
            art_bord = 110,
            allowed_percent_artpoints = 70,
            raw_betap_limit = 800_000,
            total_pow_border = (uint)(8 * 1e7),
            global_artwin_sec = 4,
            spect_art_by_totalp = false,
            hanning_win_spectrum = false,
            hamming_win_spectrum = true,
            num_wins_for_quality_avg = 100
        };

        var shortArtsDetect = new ShortArtifactDetectSetting
        {
            ampl_art_detect_win_size = 200,
            ampl_art_zerod_area = 200,
            ampl_art_extremum_border = 25
        };

        var mentalAndSpectralSettings = new MentalAndSpectralSetting
        {
            n_sec_for_instant_estimation = 2,
            n_sec_for_averaging = 4
        };
        return new EegEmotionalMath(mathLib, artsDetect, shortArtsDetect, mentalAndSpectralSettings);
    }
}

public class BrainBitController : MonoBehaviour
{
    #region Init
    void Awake()
    {
        DontDestroyOnLoad(this.gameObject);
    }
    #endregion

    #region MainThreadImpl
    private Queue<Action> jobs = new Queue<Action>();

    void Update()
    {
        while (jobs.Count > 0)
        {
            Debug.Log("Job invoked");
            jobs.Dequeue().Invoke();
        }
            
    }

    internal void AddJob(Action newJob)
    {
        jobs.Enqueue(newJob);
    }
    #endregion


    #region Fields
    private Scanner scanner;
    private Dictionary<string, BrainBitAdditional> connectedDevices = new();
    private List<string> disconnectedDevices = new();

    public IReadOnlyList<string> ConnectedDevices => connectedDevices.Keys.ToList();

    #endregion

    #region Interface
    public event Action<string, ConnectionState> EventConnectionStateChanged;
    public event Action<string, int> EventBatteryChanged;

    public async Task<IEnumerable<BrainBitInfo>> SearchWithResult(int seconds, List<string> addresses)
    {
        return await Task.Run(async () =>
        {
            createScanner();
            scanner.Start();
            await Task.Delay(seconds * 1000);
            scanner.Stop();
            var founded = scanner.Sensors;
            if (founded.Count > 0)
            {
                return founded.Where(si => addresses.Count() < 1 || addresses.Contains(si.Address))
                    .Select(si => new BrainBitInfo()
                    {
                        Name = si.Name,
                        Address = si.Address,
                        SensInfo = si,
                        ConnectionState=ConnectionState.Disconnected
                    });
            }
            return new List<BrainBitInfo>();
        });
    }

    public async Task ConnectTo(BrainBitInfo info, bool needReconnect = false)
    {
        await Task.Run(() =>
        {
            createScanner();
            SendConnectionEvent(info.Address, ConnectionState.Connection);

            try
            {
#if ANDROID
                AddJob(() =>
#endif
                {
                    if (scanner.CreateSensor(info.SensInfo) is BrainBitSensor bb)
                    {
                        bb.EventSensorStateChanged += eventSensorStateChanged;
                        bb.EventBatteryChanged += eventBatteryChanged;
                        connectedDevices.Add(info.Address, new BrainBitAdditional(needReconnect, bb));
                        SendConnectionEvent(info.Address, ConnectionState.Connected);
                    }
                    else
                    {
                        SendConnectionEvent(info.Address, ConnectionState.Error);
                    }

                }
#if ANDROID
                );
#endif
            }
            catch (Exception ex)
            {
                SendConnectionEvent(info.Address, ConnectionState.Error);
            }
        });
    }

    private void SendConnectionEvent(string address, ConnectionState state)
    {
        Debug.Log("SendConnectionEvent");
        AddJob(() => {
            Debug.Log("SendConnectionEvent from job");
            EventConnectionStateChanged?.Invoke(address, state);
        });
    }

    public async Task DisconnectFrom(string address)
    {
        await Task.Run(() =>
        {
            SendConnectionEvent(address, ConnectionState.Disconnection);
            if (disconnectedDevices.Contains(address))
            {
                disconnectedDevices.Remove(address);
                if (disconnectedDevices.Count() < 1) scanner.Stop();
            }
            var tmp = connectedDevices[address];
            connectedDevices.Remove(address);
            tmp.bb.Disconnect();
            tmp.bb.Dispose();
            tmp.bb = null;
            SendConnectionEvent(address, ConnectionState.Disconnected);
        });
    }

#endregion

    #region Resistance
    public event Action<string, ResistValues> EventResistValueReceived;
    public async void StartResist(string address)
    {
        await Task.Run(() => {
            var tmp = connectedDevices[address].bb;
            tmp.EventBrainBitResistDataRecived += eventBrainBitResistDataRecived;
            tmp.ExecCommand(SensorCommand.CommandStartResist);
        });
    }

    public async void StopResist(string address)
    {
        await Task.Run(() => {
            var tmp = connectedDevices[address].bb;
            tmp.EventBrainBitResistDataRecived -= eventBrainBitResistDataRecived;
            tmp.ExecCommand(SensorCommand.CommandStopResist);
        });
    }

    private void eventBrainBitResistDataRecived(ISensor sensor, BrainBitResistData data)
    {
        AddJob(() => {
            EventResistValueReceived?.Invoke(sensor.Address, new ResistValues()
            {
                O1 = data.O1 < 2500000 ? ResistState.Normal : ResistState.Bad,
                O2 = data.O2 < 2500000 ? ResistState.Normal : ResistState.Bad,
                T3 = data.T3 < 2500000 ? ResistState.Normal : ResistState.Bad,
                T4 = data.T4 < 2500000 ? ResistState.Normal : ResistState.Bad,
            });
        });
    }
    #endregion

    #region Internal
    private void eventBatteryChanged(ISensor sensor, int battPower)
    {
        AddJob(() =>
        {
            EventBatteryChanged?.Invoke(sensor.Address, battPower);
        });
        Debug.Log($"Power ({sensor.Address}): {battPower}");
    }

    private void eventSensorStateChanged(ISensor sensor, SensorState sensorState)
    {
        SendConnectionEvent(sensor.Address, sensorState == SensorState.StateInRange ? ConnectionState.Connected : ConnectionState.Disconnected);
        
        if (sensorState == SensorState.StateOutOfRange && connectedDevices.ContainsKey(sensor.Address) && connectedDevices[sensor.Address].needReconnect)
        {
            disconnectedDevices.Add(sensor.Address);
            AddJob(() =>
            {
                if (scanner == null) createScanner();
                scanner.Start();
            });
        }
    }

    private void eventSensorsChanged(IScanner scanner, IReadOnlyList<SensorInfo> sensors)
    {
        var founded = sensors.Where(si => disconnectedDevices.Contains(si.Address));
        if (founded.Count() > 0)
        {
            scanner.Stop();
            var bbAdditional = connectedDevices[founded.First().Address];
            bbAdditional.bb.Connect();
            if (bbAdditional.bb.State != SensorState.StateInRange)
            {
                scanner.Start();
            }
            else
            {
                if (bbAdditional.isSignal)
                {
                    bbAdditional.bb.ExecCommand(SensorCommand.CommandStartSignal);
                }
                disconnectedDevices.Remove(founded.First().Address);
            }
        }
    }

    private void createScanner()
    {
        if(scanner == null)
        {
            scanner = new Scanner(SensorFamily.SensorLEBrainBit);
            scanner.EventSensorsChanged += eventSensorsChanged;
        }
        
    }
    #endregion

    #region Calculations
    public event Action<string, MindDataReal> EventMindDataUpdated;
    public event Action<string, int> EventCalibrationProgressChanged;
    public event Action<string, bool> EventArtefactFounded;

    public async void StartCalculations(string address)
    {
        await Task.Run(() => {
            var tmp = connectedDevices[address].bb;
            connectedDevices[address].calibrationStarted = true;
            connectedDevices[address].emotionalMath.StartCalibration();
            tmp.EventBrainBitSignalDataRecived += eventBrainBitSignalDataRecived;
            tmp.ExecCommand(SensorCommand.CommandStartSignal);
            connectedDevices[address].isSignal = true;
        });
    }

    private void eventBrainBitSignalDataRecived(ISensor sensor, BrainBitSignalData[] data)
    {
        var samples = new RawChannels[data.Length];

        for (var i = 0; i < data.Length; i++)
        {
            samples[i].LeftBipolar = data[i].T3 - data[i].O1;
            samples[i].RightBipolar = data[i].T4 - data[i].O2;
        }

        var math = connectedDevices[sensor.Address].emotionalMath;

        try
        {
            math.PushData(samples);
            math.ProcessDataArr();
            AddJob(() =>
            {
                EventArtefactFounded?.Invoke(sensor.Address, math.IsBothSidesArtifacted());
            });
            Debug.Log($"Is artefacted: {math.IsBothSidesArtifacted()}");
            if (connectedDevices[sensor.Address].calibrationStarted)
            {
                if (math.CalibrationFinished())
                {
                    connectedDevices[sensor.Address].calibrationStarted = false;
                    AddJob(() => 
                    {
                        EventCalibrationProgressChanged?.Invoke(sensor.Address, 100);
                    });
                }
                else {
                    int progress = math.GetCallibrationPercents();
                    AddJob(() =>
                    {
                        EventCalibrationProgressChanged?.Invoke(sensor.Address, progress);
                    });
                }
            }
            else
            {
                MindData[] mindData = math?.ReadMentalDataArr();
                if (mindData.Length > 0)
                {
                    MindData md = mindData.Last();
                    AddJob(() => {
                        EventMindDataUpdated?.Invoke(sensor.Address, new MindDataReal()
                        {
                            Relaxation = md.RelRelaxation,
                            Attention = md.RelAttention
                        });
                    });
                }
            }
        }
        catch (Exception e)
        {
            Debug.Log(e.Message);
        }
    }

    public async void StopCalculations(string address) 
    {
        await Task.Run(() => {
            var tmp = connectedDevices[address].bb;
            connectedDevices[address].calibrationStarted = false;
            tmp.EventBrainBitSignalDataRecived -= eventBrainBitSignalDataRecived;
            tmp.ExecCommand(SensorCommand.CommandStopSignal);
            connectedDevices[address].isSignal = true;
        });
    }
    #endregion

    public void StopAll()
    {
        if (scanner != null)
        {
            scanner.EventSensorsChanged -= eventSensorsChanged;
            scanner.Stop();
            scanner.Dispose();
            scanner = null;
        }
        foreach(var bb in connectedDevices.Values)
        {
            bb.bb.Disconnect();
            bb.bb.Dispose();
        }
        connectedDevices.Clear();
    }
}

File: students/PythonDemo/brain_bit_controller.py
import contextlib
import enum
from threading import Thread
from dataclasses import dataclass
from typing import List

from PyQt6.QtCore import QObject, pyqtSignal, QThread
from em_st_artifacts.emotional_math import EmotionalMath
from em_st_artifacts.utils.lib_settings import ArtifactDetectSetting, ShortArtifactDetectSetting, \
    MentalAndSpectralSetting, MathLibSetting
from em_st_artifacts.utils.support_classes import RawChannels
from neurosdk.brainbit_sensor import BrainBitSensor
from neurosdk.scanner import Scanner
from neurosdk.sensor import Sensor
from neurosdk.cmn_types import *
from em_st_artifacts import emotional_math

class ConnectionState(Enum):
    Connection=0
    Connected=1
    Disconnection=2
    Disconnected=3
    Error=4


@dataclass
class MindDataReal:
    attention: float
    relaxation: float


class ResistState(Enum):
    Normal=0
    Bad=1


@dataclass
class ResistValues:
    O1: ResistState
    O2: ResistState
    T3: ResistState
    T4: ResistState


@dataclass
class BrainBitInfo:
    Name: str
    Address: str
    sensor_info: SensorInfo


class BrainBitAdditional:
    def __init__(self, need_reconnect, sensor):
        self.need_reconnect: bool = need_reconnect
        self.bb: BrainBitSensor = sensor
        self.is_signal = False
        self.emotional_math: emotional_math.EmotionalMath=self.__create_emotional_math()

    def __create_emotional_math(self) -> EmotionalMath:
        mls = MathLibSetting(sampling_rate=250,
                             process_win_freq=25,
                             fft_window=1000,
                             n_first_sec_skipped=4,
                             bipolar_mode=True,
                             squared_spectrum=True,
                             channels_number=4,
                             channel_for_analysis=0)

        ads = ArtifactDetectSetting(hanning_win_spectrum=True, num_wins_for_quality_avg=125)

        sads = ShortArtifactDetectSetting(ampl_art_extremum_border=25)

        mss = MentalAndSpectralSetting()
        calibration_length = 6
        nwins_skip_after_artifact = 5
        return EmotionalMath(mls, ads, sads, mss)


class Worker(QObject):
    finished = pyqtSignal()

    def __init__(self, work):
        super().__init__()
        self.work = work

    def run(self):
        self.work()
        self.finished.emit()


class BrainBitController(QObject):
    connectionStateChanged = pyqtSignal(str, ConnectionState)
    batteryChanged = pyqtSignal(str, int)
    resistValuesUpdated=pyqtSignal(str, ResistValues)
    mindDataUpdated = pyqtSignal(str, MindDataReal)
    isArtefacted = pyqtSignal(str, bool)
    calibrationProcessChanged = pyqtSignal(str, int)
    founded = pyqtSignal(list)

    def __init__(self):
        super().__init__()
        self.__calibration_started = False
        self.__scanner = Scanner([SensorFamily.LEBrainBit])
        self.__connected_devices = {}
        self.__disconnected_devices=list()
        self.connected_devices=list()
        self.thread = None
        self.worker = None

    def search_with_result(self, seconds: int, addresses: List[str]):
        def __device_scan():
            self.__scanner.start()
            QThread.sleep(seconds)
            self.__scanner.stop()
            self.founded.emit([BrainBitInfo(Name=si.Name,
                                            Address=si.Address,
                                            sensor_info=si) for si in self.__scanner.sensors()])

        self.thread = QThread()
        self.worker = Worker(__device_scan)
        self.worker.moveToThread(self.thread)
        self.thread.started.connect(self.worker.run)
        self.worker.finished.connect(self.thread.quit)
        self.thread.start()

    def connect_to(self, info: BrainBitInfo, need_reconnect: bool = False):
        self.connectionStateChanged.emit(info.Address, ConnectionState.Connection)
        def __device_connection():
            sensor=None
            try:
                sensor = self.__scanner.create_sensor(info.sensor_info)
            except Exception as err:
                print(err)
            try:
                if sensor is not None:
                    sensor.sensorStateChanged = self.__connection_state_changed
                    sensor.batteryChanged = self.__battery_changed
                    self.__connected_devices.update({info.Address: BrainBitAdditional(need_reconnect, sensor)})
                    self.connected_devices.append(info.Address)
                    self.connectionStateChanged.emit(info.Address, ConnectionState.Connected)
                else:
                    self.connectionStateChanged.emit(info.Address, ConnectionState.Error)
            except Exception as err:
                print(err)

        self.thread = QThread()
        self.worker = Worker(__device_connection)
        self.worker.moveToThread(self.thread)
        self.thread.started.connect(self.worker.run)
        self.worker.finished.connect(self.thread.quit)
        self.thread.start()

    def __event_sensor_founded(self, scanner: Scanner, sensors: List[SensorInfo]):
        scanner.stop()
        for si in sensors:
            if si.Address in self.__disconnected_devices:
                self.__connected_devices[si.Address].bb.connect()
                if self.__connected_devices[si.Address].bb.state != SensorState.StateInRange:
                    scanner.start()
                else:
                    if self.__connected_devices[si.Address].is_signal:
                        self.__execute_command(self.__connected_devices[si.Address].bb, SensorCommand.StartSignal)
                    self.__disconnected_devices.remove(si.Address)
                return
        if len(self.__disconnected_devices) > 0:
            scanner.start()

    def __connection_state_changed(self, sensor: Sensor, state: SensorState):
        self.connectionStateChanged.emit(sensor.address,
                                         ConnectionState.Connected if state == SensorState.StateInRange else ConnectionState.Disconnected)
        if state == SensorState.StateOutOfRange and sensor.address in self.__connected_devices.keys() and \
                self.__connected_devices[sensor.address].need_reconnect:
            self.__disconnected_devices.append(sensor.address)
            self.__scanner.sensorsChanged = self.__event_sensor_founded
            self.__scanner.start()

    def __battery_changed(self, sensor: Sensor, battery: int):
        self.batteryChanged.emit(sensor.address, battery)

    def disconnect_from(self, address: str):
        self.connectionStateChanged.emit(address, ConnectionState.Disconnection)
        if address in self.__disconnected_devices:
            self.__disconnected_devices.remove(address)
            if len(self.__disconnected_devices) < 1:
                self.__scanner.stop()
        sens = self.__connected_devices[address]
        self.__connected_devices.pop(address)
        self.connected_devices.remove(address)
        sens.bb.disconnect()
        sens.bb = None

    def start_resist(self, address: str):
        def on_resist_received(sensor, data):
            try:
                resistValues = ResistValues(O1=ResistState.Normal if data.O1 < 2500000 else ResistState.Bad,
                                            O2=ResistState.Normal if data.O2 < 2500000 else ResistState.Bad,
                                            T3=ResistState.Normal if data.T3 < 2500000 else ResistState.Bad,
                                            T4=ResistState.Normal if data.T4 < 2500000 else ResistState.Bad)
                self.resistValuesUpdated.emit(sensor.address, resistValues)
            except Exception as err:
                print(err)

        try:
            self.__connected_devices[address].bb.resistDataReceived = on_resist_received
            self.__execute_command(self.__connected_devices[address].bb, SensorCommand.StartResist)
        except Exception as err:
            print(err)

    def stop_resist(self, address: str):
        try:
            self.__connected_devices[address].bb.resistDataReceived=None
            self.__execute_command(self.__connected_devices[address].bb, SensorCommand.StopResist)
        except Exception as err:
            print(err)

    def start_calculations(self, address: str):
        def on_signal_received(sensor, data):
            math = self.__connected_devices[address].emotional_math

            raw_channels = []
            for sample in data:
                left_bipolar = sample.T3 - sample.O1
                right_bipolar = sample.T4 - sample.O2
                raw_channels.append(RawChannels(left_bipolar, right_bipolar))

            math.push_data(raw_channels)
            math.process_data_arr()

            self.isArtefacted.emit(address, math.is_both_sides_artifacted())

            if self.__calibration_started:
                if math.calibration_finished():
                    self.__calibration_started=False
                    self.calibrationProcessChanged.emit(address, 100)
                else:
                    self.calibrationProcessChanged.emit(address, math.get_calibration_percents())
            else:
                mental_data=math.read_mental_data_arr()
                if len(mental_data)>0:
                    md=mental_data[-1]
                    self.mindDataUpdated.emit(address, MindDataReal(attention=md.rel_attention,
                                                                         relaxation=md.rel_relaxation))

        try:
            self.__calibration_started = True
            self.__connected_devices[address].emotional_math.start_calibration()
            self.__connected_devices[address].bb.signalDataReceived = on_signal_received
            self.__execute_command(self.__connected_devices[address].bb, SensorCommand.StartSignal)
            self.__connected_devices[address].is_signal = True
        except Exception as err:
            print(err)


    def stop_calculations(self, address: str):
        self.__calibration_started = False
        self.__connected_devices[address].bb.signalDataReceived = None
        self.__execute_command(self.__connected_devices[address].bb, SensorCommand.StopSignal)
        self.__connected_devices[address].is_signal = False

    def __execute_command(self, sensor: Sensor, command: SensorCommand):
        def execute_command():
            try:
                sensor.exec_command(command)
            except Exception as err:
                print(err)
        thread = Thread(target=execute_command)
        thread.start()

    def stop_all(self):
        self.__scanner.stop()
        self.__scanner.sensorsChanged = None
        self.__scanner = None

        for device in self.__connected_devices.values():
            device.bb.disconnect()
            device.bb.sensorStateChanged = None
            device.bb.batteryChanged = None
            device.bb = None
        self.__connected_devices.clear()
        self.connected_devices.clear()
        self.__disconnected_devices.clear()

brain_bit_controller = BrainBitController()